README.md
# bedrock-consumer

[![Build Status](http://ci.digitalbazaar.com/buildStatus/icon?job=bedrock-consumer)](http://ci.digitalbazaar.com/job/bedrock-consumer)

A [bedrock][] library that provides simpler helper methods for authenticating
and establishing user sessions and a method to verify and convert a HTTP-posted
identity profile to an event-based API to assist implementers of custom
processing of credentials. It is not a comprehensive credential
consumer application, rather it provides some basic, common capabilities for
applications that consume credentials.

This module:

* Provides [express][] middleware for decentralized identity authentication,
  creating and destroying sessions with an application, and converting
  a posted identity profile that contains credentials into an event-based
  API for processing those credentials
* Attaches some user information to session state to assist frontend
  applications
* Works in conjunction with [bedrock-passport][] to setup default user
  settings and permissions for non-persistent users that log into an
  application

## Requirements

- npm v3+

## Middleware

This module provides some express middleware:

* **login** - This middleware will attempt to establish an authenticated
  session for the user that sent the request. Upon success, a status code
  of 200 and the identity that was authenticated are sent to the client. If
  the user can't be authenticated, a status code of 400 is returned. See
  [bedrock-passport][] or [bedrock-identity][] for more information.
* **logout** - This middleware will destroy any currently authenticated
  session and return a status code of 204.
* **authenticate** - This middleware will attempt to authenticate the given
  decentralized identity by verifying its digital signature and any signatures
  found on its associated credentials. A status code of 200 and a response
  body including the authenticated identity and verified credentials is
  returned to the client on success. If the user cannot be authenticated, a
  status code of 400 is returned.
* **handleCredential** - This middleware will attempt to verify the posted
  identity profile containg credentials, and on success, will pass them to
  [bedrock][]'s event API using the custom
  `bedrock-consumer.credentials.process` event. Any event handlers will be
  passed event data with the [express] `request`, `response`, and `next`
  middleware handler, as well as the `identity` and `verifyResults`. Once
  all event handlers have executed (or once one of them cancels and takes over
  the process), a status code of 204 will be returned to the client.

[bedrock]: https://github.com/digitalbazaar/bedrock
[bedrock-identity]: https://github.com/digitalbazaar/bedrock-identity
[bedrock-passport]: https://github.com/digitalbazaar/bedrock-passport
[express]: https://github.com/expressjs/express


CHANGELOG.md
# bedrock-consumer ChangeLog

## 2.0.6 - 2016-12-21

### Fixed
- Ensure identity defaults are copied before modification.

## 2.0.5 - 2016-11-10

### Changed
- Use new `generateResource` and `exists` APIs.

## 2.0.4 - 2016-08-03

### Fixed
- Fix typo in the bedrock-passport.authenticate handler.

## 2.0.3 - 2016-04-26

### Changed
- Update dependencies.

## 2.0.2 - 2016-04-16

### Fixed
- Remove peer deps from dependencies list.

## 2.0.1 - 2016-03-24

### Changed
- Update bedrock-session-http references.

## 2.0.0 - 2016-03-03

### Changed
- Update package dependencies for npm v3 compatibility.

## 1.0.0 - 2016-01-31

- See git history for changes.


test/test.js
/*!
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
var bedrock = require('bedrock');
require('bedrock-consumer');

require('bedrock-test');
bedrock.start();


test/test.config.js
/*!
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
var config = require('bedrock').config;
var path = require('path');

config.mocha.tests.push(path.join(__dirname, 'mocha'));


test/mocha/10-api.js
/*!
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
/* globals should */
var bedrock = require('bedrock');
var mockData = require('./mock.data');

describe('bedrock-consumer', () => {
  it('authenticate handler adds identity defaults', done => {
    var authInfo = bedrock.util.clone(mockData.authenticateInfo);
    bedrock.config.consumer.defaults.identity.sysResourceRole = [{
      sysRole: 'd95bcf89-eca7-41d3-a8bb-af5b8bd78e7d',
      generateResource: 'id'
    }];
    bedrock.events.emit('bedrock-passport.authenticate', authInfo, () => {
      var i = authInfo.user.identity;
      should.exist(i.sysResourceRole);
      i.sysResourceRole.should.be.an('array');
      i.sysResourceRole.should.have.length(1);
      var role = i.sysResourceRole[0];
      should.not.exist(role.generateResource);
      role.should.be.an('object');
      should.exist(role.sysRole);
      role.sysRole.should.equal('d95bcf89-eca7-41d3-a8bb-af5b8bd78e7d');
      should.exist(role.resource);
      role.resource.should.be.an('array');
      role.resource.should.have.length(1);
      role.resource[0].should.equal(authInfo.user.identity.id);
      done();
    });
  });
});


lib/consumer.js
/*
 * Bedrock Consumer module.
 *
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var brIdentity = require('bedrock-identity');
// FIXME: REMOVE
// var brPassport = require('bedrock-passport');
// var BedrockError = bedrock.util.BedrockError;
var cio = require('credentials-io');

// load config defaults
require('./config');

// module API
var api = {};
module.exports = api;

// TODO: configure `cio` to use `bedrock.jsonld`

/**
 * Process an Identity containing credentials by verifying the basic
 * credential information and then emitting an event containing the
 * the verified credential information and all the data necessary to
 * respond to the request. Credential consumer event listeners will
 * determine how to consume the given credentials and then complete the
 * response to the request via the res/next parameters.
 *
 * Emits the following event when receiving a credential:
 *
 * Event: bedrock-consumer.credentials.process
 * Data:
 *   request the HTTP request.
 *   response the HTTP response.
 *   next the next callback in the express chain.
 *   identity the verified identity.
 *   verifyResults the results of the verification process.
 *
 * @param req the HTTP request.
 * @param res the HTTP response.
 * @param next the next callback in the express chain.
 */
api.handleCredential = function(req, res, next) {
  async.auto({
    verify: function(callback) {
      // TODO: enable timestamp checking on sandbox
      var checkTimestamp = !(bedrock.config.environment === 'development' ||
        bedrock.config.environment === 'testing' ||
        bedrock.config.environment === 'sandbox');
      cio.verify(req.body, {
        checkTimestamp: checkTimestamp
      }, callback);
    },
    emit: ['verify', function(callback, results) {
      bedrock.events.emit(
        'bedrock-consumer.credentials.process', {
          request: req,
          response: res,
          next: next,
          identity: req.body,
          verifyResults: results.verify
        }, callback);
    }]
  }, function(err, results) {
    if(results.emit === false) {
      // the event, along with the HTTP response, was handled by another module
      return;
    }
    if(err) {
      // there was an error while emitting the event
      return next(err);
    }

    // the event was not handled by another module, handle the HTTP response
    res.sendStatus(204);
  });
};

/**
 * Attempt to authenticate the given identity. Upon success, a status code of
 * 200 and the identity that was authenticated and any related authentication
 * information are sent to the client.
 *
 * @param req the request.
 * @param res the response.
 * @param next the next route handler.
 */
// FIXME: it appears that this API is not used anywhere, remove if not needed
/*
api.authenticate = function(req, res, next) {
  brPassport.authenticate('did', function(err, user) {
    if(!user) {
      // user not authenticated
      err = new BedrockError(
        'The given decentralized identity could not be authenticated.',
        'NotAuthenticated', {'public': true, httpStatusCode: 400}, err);
    }
    if(err) {
      return next(err);
    }
    res.status(200).send(
      {identity: user.identity, credentials: user.credentials});
  })(req, res, next);
};

*/

// add identity defaults to non-persistent identities
bedrock.events.on('bedrock-passport.authenticate', function(info, callback) {
  var config = bedrock.config.consumer;
  var identity = info.user.identity;
  async.auto({
    isPersistent: function(callback) {
      // TODO: optimize to existence check or check to see if strategy already
      // did the check
      brIdentity.get(null, identity.id, function(err) {
        if(err) {
          if(err.name === 'NotFound') {
            return callback(null, false);
          }
          return callback(err);
        }
        callback(null, true);
      });
    },
    addDefaults: ['isPersistent', function(callback, results) {
      // do not add defaults for a persistent user, any defaults would
      // be handled by `bedrock-identity` instead
      if(results.isPersistent) {
        return callback();
      }
      // apply defaults to identity
      info.user.identity = _.assign({}, config.defaults.identity, identity);

      // handle `generateResource` feature
      var roles = bedrock.jsonld.getValues(identity, 'sysResourceRole');
      roles.forEach(function(role) {
        if(role.generateResource === 'id') {
          role.resource = [identity.id];
        }
      });
      callback();
    }]
  }, function(err) {
    callback(err);
  });
});

// FIXME: remove if not needed
/*
// optionally install login handler
var brExpress;
try {
  brExpress = require('bedrock-express');
} catch(e) {}

if(brExpress) {
  bedrock.events.on('bedrock-express.configure.routes', function(app) {
    var routes = bedrock.config.consumer.routes;
    app.post(routes.login, api.login);
    app.post(routes.authenticate, api.authenticate);
    app.get(routes.logout, api.logout);
  });
}
*/

// optionally install client-side session access
var brSession;
try {
  brSession = require('bedrock-session-http');
} catch(e) {}

if(brSession) {
  bedrock.events.on('bedrock-session-http.session.get', function(req, session) {
    if(req.isAuthenticated() && req.user.identity && req.user.identity.id) {
      session.identity = session.identity || {};
      session.identity.id = req.user.identity.id;
    }
  });
}


lib/index.js
/*
 * Bedrock Consumer module.
 *
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var brIdentity = require('bedrock-identity');
var brPassport = require('bedrock-passport');
var cio = require('credentials-io');
var BedrockError = bedrock.util.BedrockError;

// load config defaults
require('./config');

// module API
var api = {};
module.exports = api;

// TODO: configure `cio` to use `bedrock.jsonld`

/**
 * Process an Identity containing credentials by verifying the basic
 * credential information and then emitting an event containing the
 * the verified credential information and all the data necessary to
 * respond to the request. Credential consumer event listeners will
 * determine how to consume the given credentials and then complete the
 * response to the request via the res/next parameters.
 *
 * Emits the following event when receiving a credential:
 *
 * Event: bedrock-consumer.credentials.process
 * Data:
 *   request the HTTP request.
 *   response the HTTP response.
 *   next the next callback in the express chain.
 *   identity the verified identity.
 *   verifyResults the results of the verification process.
 *
 * @param req the HTTP request.
 * @param res the HTTP response.
 * @param next the next callback in the express chain.
 */
api.handleCredential = function(req, res, next) {
  async.auto({
    verify: function(callback) {
      // TODO: enable timestamp checking on sandbox
      var checkTimestamp = !(bedrock.config.environment === 'development' ||
        bedrock.config.environment === 'testing' ||
        bedrock.config.environment === 'sandbox');
      cio.verify(req.body, {
        checkTimestamp: checkTimestamp
      }, callback);
    },
    emit: ['verify', function(callback, results) {
      bedrock.events.emit(
        'bedrock-consumer.credentials.process', {
          request: req,
          response: res,
          next: next,
          identity: req.body,
          verifyResults: results.verify
        }, callback);
    }]
  }, function(err, results) {
    if(results.emit === false) {
      // the event, along with the HTTP response, was handled by another module
      return;
    }
    if(err) {
      // there was an error while emitting the event
      return next(err);
    }

    // the event was not handled by another module, handle the HTTP response
    res.sendStatus(204);
  });
};

/**
 * Attempt to establish an authorized session for the user that sent the
 * request. Upon success, a status code of 200 and the identity that was
 * authenticated are sent to the client.
 *
 * @param req the request.
 * @param res the response.
 * @param next the next route handler.
 */
api.login = function(req, res, next) {
  brPassport.authenticate('did', function(err, user) {
    if(!user) {
      // user not authenticated
      err = new BedrockError(
        'The given decentralized identity could not be authenticated.',
        'NotAuthenticated', {'public': true, httpStatusCode: 400}, err);
    }
    if(err) {
      return next(err);
    }
    req.login(user, function(err) {
      if(err) {
        return next(err);
      }
      res.status(200).send(user.identity);
    });
  })(req, res, next);
};

api.logout = function(req, res, next) {
  req.logout();
  res.sendStatus(204);
};

/**
 * Attempt to authenticate the given identity. Upon success, a status code of
 * 200 and the identity that was authenticated and any related authentication
 * information are sent to the client.
 *
 * @param req the request.
 * @param res the response.
 * @param next the next route handler.
 */
api.authenticate = function(req, res, next) {
  brPassport.authenticate('did', function(err, user) {
    if(!user) {
      // user not authenticated
      err = new BedrockError(
        'The given decentralized identity could not be authenticated.',
        'NotAuthenticated', {'public': true, httpStatusCode: 400}, err);
    }
    if(err) {
      return next(err);
    }
    res.status(200).send(
      {identity: user.identity, credentials: user.credentials});
  })(req, res, next);
};

// add identity defaults to non-persistent identities
bedrock.events.on('bedrock-passport.authenticate', (info, callback) => {
  var config = bedrock.config.consumer;
  var identity = info.user.identity;
  async.auto({
    isPersistent: callback => brIdentity.exists(null, identity.id, callback),
    addDefaults: ['isPersistent', (callback, results) => {
      // do not add defaults for a persistent user, any defaults would
      // be handled by `bedrock-identity` instead
      if(results.isPersistent) {
        return callback();
      }
      // apply defaults to identity
      info.user.identity = _.assign(
        {}, _.cloneDeep(config.defaults.identity), identity);

      // handle `generateResource` feature
      var roles =
        bedrock.jsonld.getValues(info.user.identity, 'sysResourceRole');
      for(var i = 0; i < roles.length; ++i) {
        if(roles[i].generateResource === 'id') {
          roles[i] = brIdentity.generateResource(roles[i], identity.id);
        }
      }
      callback();
    }]
  }, err => callback(err));
});

// optionally install login handler
var brExpress;
try {
  brExpress = require('bedrock-express');
} catch(e) {}

if(brExpress) {
  bedrock.events.on('bedrock-express.configure.routes', function(app) {
    var routes = bedrock.config.consumer.routes;
    app.post(routes.login, api.login);
    app.post(routes.authenticate, api.authenticate);
    app.get(routes.logout, api.logout);
  });
}

// optionally install client-side session access
var brSession;
try {
  brSession = require('bedrock-session-http');
} catch(e) {}

if(brSession) {
  bedrock.events.on('bedrock-session-http.session.get', function(req, session) {
    if(req.isAuthenticated() && req.user && req.user.identity &&
      req.user.identity.id) {
      session.identity = session.identity || {};
      session.identity.id = req.user.identity.id;
    }
  });
}
